### PR TITLE
ci: smoke test with/out optional dependencies, mark more tests

### LINKED ISSUE #1905: Testing without optional dependencies
Bugs may occur when optional packages are either installed or not installed. These issues may evade the current CI, which assumes all optional dependencies are installed. I'm wondering if it is worthwhile to run a few tests in serial. For example, here is a sequence a worker would do:
```bash
# basic install without any optional dependencies
pip install .[test]
pytest -S --cov
# re-install with all optional dependencies
pip install .[optional]
pytest -S --cov --cov-append
# remove a few optional dependencies
pip uninstall --yes pyproj shapely [possibly a few others]
pytest -S --cov --cov-append
```
note that the coverage reports are appended, which could potentially increase the coverage to cover logic that assumes optional decencies are installed or not.

### PR DESCRIPTION
**Pull request by w-bonelli**
- close #1905
  - run smoke tests with all, some and no optional dependencies on nightly schedule
  - in partial dependency availability case, remove 3 randomly chosen dependencies every day
- mark some tests requiring optional dependencies
- only check crs deprecations if pyproj available

### CODE REVIEW

**Review by mwtoews — COMMENTED** [review]
Looks good so far. It adds about 5 mins to the smoke tests. It seems the smoke test does not upload it's coverage, it probably should to see the changes.

File: autotest/test_grid.py [hunk 1]
```diff
@@ -675,9 +679,12 @@ def do_checks(g, *, exp_srs=expected_srs, exp_epsg=expected_epsg):
     sg.set_coord_info(crs=26915, merge_coord_info=False)
     do_checks(sg, exp_srs="EPSG:26915", exp_epsg=26915)
 
+    # only check deprecations if pyproj is available
+    context = pytest.deprecated_call() if HAS_PYPROJ else nullcontext()
```
[the remark this answers is not in the document]
Comment (w-bonelli) [reply autotest/test_grid.py]: renamed to `pyproj_avail_context`

**Comment by w-bonelli** [comment]
@mwtoews I wonder if it may be worth randomizing the choice of dependencies to test without. I added a mechanism for that here, and disabled the coverage upload for now. Let me know if this seems unreasonable. My thinking is tests can be expanded if/when issues come to light &mdash; later if we are more confident optional code paths are thoroughly/deterministically tested we could drop the random tests and enable smoke test coverage reporting.

**Comment by mwtoews** [comment]
I like the idea to randomize optional deps, as there are potential combinations that may cause issues. We can see how it goes. If it becomes too much of a maintenance burden, then the randomization can be disabled. And I'd agree it doesn't make sense to upload an unstable coverage report.

File: .github/workflows/commit.yml [hunk 2]
```diff
@@ -114,33 +116,43 @@ jobs:
...
           pip install --upgrade pip
           pip install .
-          pip install ".[test, optional]"
+          pip install ".[test,optional]"
+      
+      - name: Upload removed dependencies log
+        uses: actions/upload-artifact@v3
+        with:
+          name: smoke-test-removed-dependencies
+          path: ./removed_dependencies.txt
```
Comment (mwtoews) [inline .github/workflows/commit.yml:133]: This doesn't seem to be needed anymore, and it shows a warning.

**Review by mwtoews — COMMENTED** [review]
It's probably better to keep the smoke tests short, so moving optional tests to their own workflow seems the right way. However, I don't see any tests from optional.yml.

File: .github/workflows/commit.yml (hunk 2 above)
Comment (w-bonelli) [reply .github/workflows/commit.yml]: thanks, removed.

**Comment by w-bonelli** [comment]
I put optional.yml on a nightly schedule and set to choose a different random 3 dependencies every day, hopefully that should not be too burdensome to stay on top of and we also avoid bloating regular CI.

**Comment by mwtoews** [comment]
That's why it's not triggered with pull requests. A schedule should work well.

**Review by mwtoews — APPROVED** [review]
Thanks!

**Pull request merged**